### PR TITLE
[FLINK-27696] Add bin-pack strategy to split the whole bucket data files into several small splits

### PR DESCRIPTION
**Pull request by JingsongLi**
We don't have to assign each task with a whole bucket data files. Instead, we can use some algorithm ( such as bin-packing) to split the whole bucket data files into multiple fragments to improve the job parallelism.

For merge tree table:
Suppose now there are files: [1, 2] [3, 4] [5, 180] [5, 190] [200, 600] [210, 700]
Files without intersection are not related, we do not need to put all files into one split, we can slice into multiple splits, multiple parallelism execution is faster. Nor can we slice too fine, we should make each split as large as possible with 128 MB, so use BinPack to slice, the final result will be:

- split1: [1, 2] [3, 4]
- split2: [5, 180] [5, 190]
- split3: [200, 600] [210, 700]

### CODE REVIEW

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/KeyValueFileStore.java [hunk 1]
```diff
@@ -93,4 +93,8 @@ public KeyValueFileStoreWrite newWrite() {
                 newScan(),
                 options.mergeTreeOptions());
     }
+
+    public Comparator<RowData> newKeyComparator() {
```
Comment (LadyForest) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/KeyValueFileStore.java:97]: Nit: replace L#75 with `newKeyComparator()`?

File: flink-table-store-connector/src/test/java/org/apache/flink/table/store/connector/source/FileStoreSourceSplitGeneratorTest.java [hunk 2]
```diff
@@ -77,12 +77,12 @@ public List<ManifestEntry> files() {
...
                 };
-        TableScan.Plan tableScanPlan =
-                new TableScan.Plan(
-                        1L,
-                        new DefaultSplitGenerator(
-                                        new FileStorePathFactory(new Path(tempDir.toString())))
-                                .generate(plan.groupByPartFiles()));
+        List<Split> scanSplits =
+                TableScan.generateSplits(
+                        new FileStorePathFactory(new Path("/tmp/")),
```
Comment (LadyForest) [inline flink-table-store-connector/src/test/java/org/apache/flink/table/store/connector/source/FileStoreSourceSplitGeneratorTest.java:82]: Why not use `tempDir`?

**Review by LadyForest — COMMENTED** [review]
@JingsongLi thanks for the contribution. It looks good in general and I just left some minor comments.

**Comment by LadyForest** [comment]
Thanks for the update and it looks good to me once the conflicts are resolved.

**Comment by JingsongLi** [comment]
@LadyForest Thanks for you review, merged.

**Pull request merged**